### PR TITLE
Prevent duplicate separators for all context menu items

### PR DESCRIPTION
**Pull request by bsclifton**
- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).

Includes all content PLUS a fix for a bug found in the original PR https://github.com/brave/browser-laptop/pull/5846
(which was reverted with https://github.com/brave/browser-laptop/pull/5867)

This PR fixes https://github.com/brave/browser-laptop/issues/5765

Auditors: @bbondy

## Includes these changes
- Added new `sanitizeTemplateItems` method to filter out bad entries
- Moved `menuUtil` from `app/browser` => `app/common`
- Moved some tests to match the new directory structure (ex: `test/unit/app`, instead of `test/unit/lib`)
- Rename template functions in `menuUtil` and items in `contextMenu.js` to reinforce distinction between **template** items and **electron  menu** items (the compiled template)
- **The new sanitize method is called in each method which builds a context menu**

## Automated test plan
- `npm run uitest -- --grep="bookmarksToolbar"` (this was the one which had a failing test which I missed)
- `npm run unittest -- --grep="sanitizeTemplateItems"`

## Manual test plan
- Launch Brave on Windows or Linux
- Right click on each of the about pages and ensure there are no double separators (might be easy to go to about:about to launch each)
- Try other context menus
  - Right clicking on a tab
  - right clicking on page
  - right clicking a bookmark / bookmark folder in the bookmarks toolbar
  - right clicking inside the downloads bar (shown after downloading an item)

### CODE REVIEW

**Comment by bsclifton** [comment]
Gonna check back in here tomorrow- hoping Travis CI loves the tests this go-around :smile:

File: test/unit/app/common/lib/menuUtilTest.js [hunk 1]
```diff
@@ -224,11 +228,44 @@ describe('menuUtil', function () {
...
+      const expectedResult = [{label: 'lol'}]
+      assert.deepEqual(result, expectedResult)
+    })
+    it('removes items which have non-string values for label or type', function () {
+      const template = [{label: true}, {type: function () { console.log('test') }}, {label: 'lol'}]
+      const result = menuUtil.sanitizeTemplateItems(template)
+      const expectedResult = [{label: 'lol'}]
+      assert.deepEqual(result, expectedResult)
+    })
+    it('always returns an array (even for one item)', function () {
```
Comment (bsclifton) [inline test/unit/app/common/lib/menuUtilTest.js]: This last test here is the one I added which failed (revealing the bug). The Array.reduce that I was doing would returning a non-array when only 1 entry exists (it now properly returns an array)

File: app/common/lib/menuUtil.js [hunk 2]
```diff
@@ -141,3 +144,40 @@ module.exports.createRecentlyClosedMenuItems = (lastClosedFrames) => {
...
+    if (isItemValid(currentValue, previousItem)) {
+      result.push(currentValue)
+    }
+    return result
+  })
+
+  return Array.isArray(result)
+    ? result
+    : [result]
+}
```
Comment (bsclifton) [inline app/common/lib/menuUtil.js]: Here is the fix for the bug which led to the revert (which happened with https://github.com/brave/browser-laptop/pull/5867)

**Comment by bbondy** [comment]
thanks!

**Comment by luixxiul** [comment]
Cleared the label and the milestone as this has been superseded with #6374

**Pull request merged**